test/net/sourceforge/kolmafia/persistence/FactDatabaseTest.java

```diff
@@ -0,0 +1,177 @@
+package net.sourceforge.kolmafia.persistence;
+
+import static internal.helpers.Player.withProperty;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.blankString;
+import static org.hamcrest.Matchers.is;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import net.sourceforge.kolmafia.AscensionClass;
+import net.sourceforge.kolmafia.AscensionPath.Path;
+import net.sourceforge.kolmafia.KoLCharacter;
+import net.sourceforge.kolmafia.RequestLogger;
+import net.sourceforge.kolmafia.objectpool.EffectPool;
+import net.sourceforge.kolmafia.persistence.FactDatabase.FactType;
+import net.sourceforge.kolmafia.preferences.Preferences;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+class FactDatabaseTest {
+  @BeforeAll
+  public static void beforeAll() {
+    KoLCharacter.reset("FactDatabaseTest");
+    Preferences.reset("FactDatabaseTest");
+  }
+
+  @Test
+  void databaseParsesWithNoErrors() {
+    var stream = new ByteArrayOutputStream();
+    try (var out = new PrintStream(stream, true)) {
+      RequestLogger.openCustom(out);
+      FactDatabase.reset();
+      RequestLogger.closeCustom();
+    }
+    String logged = stream.toString().trim();
+    assertThat(logged, blankString());
+  }
+
+  @ParameterizedTest
+  @CsvSource({
+    "SEAL_CLUBBER, WILDFIRE, ratbat, EFFECT, Egg-cellent Vocabulary (10)",
+    "SEAL_CLUBBER, NONE, Octorok, EFFECT, Egg-stortionary Tactics (10)",
+    "PASTAMANCER, NONE, Jacob's adder, EFFECT, Egg-stortionary Tactics (10)",
+    "PASTAMANCER, NONE, Black Crayon Penguin, MEAT, 149 Meat",
+    "DISCO_BANDIT, SMALL, Bob Racecar, EFFECT, Feeling Excited (15)"
+  })
+  void picksCorrectFact(
+      final AscensionClass ascensionClass,
+      final Path path,
+      final String monsterName,
+      final FactType factType,
+      final String factString) {
+    var monster = MonsterDatabase.findMonster(monsterName);
+    var fact = FactDatabase.getFact(ascensionClass, path, monster, false);
+    assertThat(fact.getType(), is(factType));
+    assertThat(fact.toString(), is(factString));
+  }
+
+  @ParameterizedTest
+  @CsvSource({
+    "0, true",
+    "1, true",
+    "2, true",
+    "3, false",
+  })
+  void pocketWishesAreStateful(final int dropped, final boolean drops) {
```

Comment:
I feel like this is a bit of a surprising way for the function to behave: you query about a monster in a different path, and Mafia tells you it drops nothing because you've already had all your pocket wish drops for today?

Is there a way to see that this would drop a pocket wish other than by clearing the pref temporarily?

Reply:
Nah that's not how this works, I'll document it in the mafia wiki once it's merged.

The runtime library functions are always reported ignoring your preference state. The monster proxy data is always reported for your current class/path/prefs.

Reply:
great, can you add a test for runtime library functions not being stateful?